libraries/standard/http/utest/http_utest.c

```diff
@@ -248,4 +341,289 @@ void test_Http_InitializeRequestHeaders_insufficient_memory()
+                                         0 /* rageEnd */ );
+    TEST_ASSERT_EQUAL( HTTP_INVALID_PARAMETER, retCode );
+
+    /* Underlying buffer is NULL in request headers. */
+    tearDown();
+    retCode = HTTPClient_AddRangeHeader( &testHeaders,
+                                         0 /* rangeStart */,
+                                         0 /* rageEnd */ );
+    TEST_ASSERT_EQUAL( HTTP_INVALID_PARAMETER, retCode );
+
+    /* Request Header Size is zero. */
+    tearDown();
+    testHeaders.pBuffer = &testBuffer[ 0 ];
+    retCode = HTTPClient_AddRangeHeader( &testHeaders,
+                                         0 /* rangeStart */,
+                                         0 /* rageEnd */ );
+    TEST_ASSERT_EQUAL( retCode, HTTP_INSUFFICIENT_MEMORY );
+
+    /* Test incorrect combinations of rangeStart and rangeEnd. */
+
+    /* rangeStart > rangeEnd */
+    tearDown();
+    testHeaders.pBuffer = &testBuffer[ 0 ];
+    retCode = HTTPClient_AddRangeHeader( &testHeaders,
+                                         10 /* rangeStart */,
+                                         5 /* rageEnd */ );
+    TEST_ASSERT_EQUAL( HTTP_INVALID_PARAMETER, retCode );
+
+    /* rangeStart is negative but rangeStart is non-End of File. */
+    tearDown();
+    testHeaders.pBuffer = &testBuffer[ 0 ];
+    retCode = HTTPClient_AddRangeHeader( &testHeaders,
+                                         -10 /* rangeStart */,
+                                         HTTP_RANGE_REQUEST_END_OF_FILE + 1 /* rageEnd */ );
+    TEST_ASSERT_EQUAL( HTTP_INVALID_PARAMETER, retCode );
+    tearDown();
+    testHeaders.pBuffer = &testBuffer[ 0 ];
+    retCode = HTTPClient_AddRangeHeader( &testHeaders,
+                                         -50 /* rangeStart */,
+                                         -10 /* rageEnd */ );
+    TEST_ASSERT_EQUAL( HTTP_INVALID_PARAMETER, retCode );
+}
+
+/**
+ * @brief Test Insufficient memory failure when the buffer has one less byte than required.
+ */
+void test_Http_AddRangeHeader_Insufficient_Memory( void )
+{
+    setupBuffersWithPreexistingHeader( &testHeaders,
+                                       testBuffer,
+                                       sizeof( testBuffer ),
+                                       &expectedHeaders,
+                                       PREEXISTING_HEADER_DATA );
+    size_t preHeadersLen = testHeaders.headersLen;
+    testRangeStart = 5;
+    testRangeEnd = 10;
+    addRangeToExpectedHeaders( &expectedHeaders,
+                               "5-10" /*expected range*/,
+                               true );
+
+    /* Update headers buffer size to be one byte short of required size to add
+     * Range Request header. */
+    testHeaders.bufferLen = expectedHeaders.dataLen - 1;
+
+    /* Re-write the expected headers buffer to store a copy of the test headers
+     * to use for verification later. */
+    memcpy( expectedHeaders.buffer, testHeaders.pBuffer, testHeaders.bufferLen );
```

Comment:
Did the helper function already do this?

Reply:
The `addRangeToExpectedHeaders` helper function added the range string to this buffer, but that data is not required in this test case as this test is expected to fail. (The `addRangeToExpectedHeaders` helper is only called to determine the total size required to hold the request data in the buffer, so that we can reduce the input buffer's length by one short of the needed total size for this test).

The data of the expected buffer is overwritten with the input buffer's to store a cache of the input buffer data to verify against data corruption by the API, which is expected to fail in this test.